packages/swing-store/src/snapStore.js

```diff
@@ -356,19 +401,129 @@ export function makeSnapStore(
     sqlDeleteSnapshotByHash.run(vatID, hash);
   }
 
-  const sqlDumpActiveSnapshots = db.prepare(`
+  const sqlGetSnapshotMetadata = db.prepare(`
+    SELECT vatID, endPos, hash, uncompressedSize, compressedSize
+    FROM snapshots
+    WHERE inUse = ?
+    ORDER BY vatID, endPos
+  `);
+
+  /**
+   * @param {boolean} includeHistorical
+   * @yields {[key: string, value: string]}
+   * @returns {Iterable<[key: string, value: string]>}
+   */
+  function* getExportRecords(includeHistorical) {
+    for (const rec of sqlGetSnapshotMetadata.iterate(1)) {
+      const exportRec = {
+        vatID: rec.vatID,
+        endPos: rec.endPos,
+        hash: rec.hash,
+        size: rec.uncompressedSize,
+      };
+      yield [artifactName(rec), JSON.stringify(exportRec)];
+      yield [`export.lastSnapshot.${rec.vatID}`, `${rec.endPos}`];
+    }
+    if (includeHistorical) {
+      for (const rec of sqlGetSnapshotMetadata.iterate(0)) {
+        const exportRec = {
+          vatID: rec.vatID,
+          endPos: rec.endPos,
+          hash: rec.hash,
+          size: rec.uncompressedSize,
+        };
+        yield [artifactName(rec), JSON.stringify(exportRec)];
+      }
+    }
+  }
+
+  function* getArtifactNames(includeHistorical) {
+    for (const rec of sqlGetSnapshotMetadata.iterate(1)) {
+      yield artifactName(rec);
+    }
+    if (includeHistorical) {
+      for (const rec of sqlGetSnapshotMetadata.iterate(0)) {
+        yield artifactName(rec);
+      }
+    }
+  }
+
+  /**
+   * @param {string} name
+   * @param {SwingStoreExporter} exporter
+   * @param {Map} metadata
+   * @returns {Promise<void>}
+   */
+  async function importSnapshot(name, exporter, metadata) {
+    const info = JSON.parse(metadata.get(name));
+    const parts = name.split('.');
+    const [tag, type, vatID, rawEndPos] = parts;
+    assert(
+      parts.length === 4 && tag === 'export' && type === 'snapshot',
+      `expected snapshot name of the form 'export.snapshot.{vatID}.{endPos}', saw '${name}'`,
+    );
+    assert(
+      info.vatID === vatID,
+      `snapshot name says vatID ${vatID}, metadata says ${info.vatID}`,
+    );
+    const endPos = Number(rawEndPos);
+    assert(
+      info.endPos === endPos,
+      `snapshot name says endPos ${endPos}, metadata says ${info.endPos}`,
+    );
+
+    const artifactChunks = exporter.getArtifact(name);
+    const inStream = Readable.from(artifactChunks);
+    let size = 0;
+    inStream.on('data', chunk => (size += chunk.length));
+    const hashStream = createHash('sha256');
+    const gzip = createGzip();
+    inStream.pipe(hashStream);
+    const compressedArtifact = await buffer(inStream.pipe(gzip));
```

Comment:
Small nit on readability and `await` hazard: all `pipe` and `.on('message')` calls need to happen synchronously (in the same turn). Being on the RHS of an `await` it's not obvious to the casual reader that the expression is indeed evaluated before the `await`.

```suggestion
    inStream.pipe(gzip);
    const compressedArtifact = await buffer(gzip);
```

Reply:
Good catch.  This dovetails with my recurring complaint about Node streams being extra confusing because the two ends of a stream are not separate objects, so when you reference a stream in your code it's ambiguous in the code text which end you're intending to refer to.